llama-index-core/llama_index/core/workflow/decorators.py

```diff
@@ -65,6 +58,13 @@ def decorator(func: Callable) -> Callable:
             requested_services=spec.requested_services or [],
         )
 
+        # If this is a free function, call add_step() explicitly.
+        if is_free_function(func.__qualname__):
+            if workflow is None:
+                msg = f"To decorate {func.__name__} please pass a workflow class to the @step decorator."
+                raise WorkflowValidationError(msg)
+            workflow.add_step(func)
+
```

Comment:
Pushed this back so that `add_step` can rely on the fact `__step_config` is set (might still fail for other validation problems)